### PR TITLE
Raise TypeError for inequalities involving zoo

### LINKED ISSUE #9037: inequalities involving zoo should raise TypeError
Since `I > 0` raises a TypeError, `zoo > 0` should probably do the same. Actually, anything even *containing* `zoo` should probably be disallowed since such an expression will become `zoo` (e.g. `zoo*1`) or `NaN` (e.g. `zoo*0`).  See #7834 for an example of how to make the changes.

### PR DESCRIPTION
**Pull request by aktech**
fixes #9037

### CODE REVIEW

**Comment by smichr** [comment]
Maybe `if me.is_complex and me.is_real is False:` could just be changed to `if me.is_complex and me.is_real is False or me is S.ComplexInfinity:`.

**Comment by aktech** [comment]
Yes, that would be better.
Thanks.

**Comment by aktech** [comment]
This Example is failing:
in [sympy/integrals/transforms.py](https://github.com/sympy/sympy/blob/master/sympy/integrals/transforms.py#L881)
```python
>>> from sympy.integrals.transforms import _simplifyconds as simp
>>> from sympy.abc import x
>>> from sympy import sympify as S
>>> simp(abs(x**2) < 1, x, 0)
Abs(x**2) < 1
```
I am not sure about the cause.

**Comment by smichr** [comment]
> I am not sure about the cause.

```
File "sympy\integrals\transforms.py", line 919, in bigger
  return bigger(1/ex2, 1/ex1)
```

When ex* is 0 this test is going to fail because a zoo will be generated.

**Comment by aktech** [comment]
@smichr  Thanks, for reply.
I should probably Except that error, using try - except.

**Comment by smichr** [comment]
The following passes tests, too. What do you think about this:

```diff
diff --git a/sympy/core/expr.py b/sympy/core/expr.py
index f3a733f..42c9960 100644
--- a/sympy/core/expr.py
+++ b/sympy/core/expr.py
@@ -235,7 +235,8 @@ def __ge__(self, other):
         except SympifyError:
             raise TypeError("Invalid comparison %s >= %s" % (self, other))
         for me in (self, other):
-            if me.is_complex and me.is_real is False:
+            if (me.is_complex and me.is_real is False) or \
+                    me.has(S.ComplexInfinity):
                 raise TypeError("Invalid comparison of complex %s" % me)
             if me is S.NaN:
                 raise TypeError("Invalid NaN comparison")
@@ -252,7 +253,8 @@ def __le__(self, other):
         except SympifyError:
             raise TypeError("Invalid comparison %s <= %s" % (self, other))
         for me in (self, other):
-            if me.is_complex and me.is_real is False:
+            if (me.is_complex and me.is_real is False) or \
+                    me.has(S.ComplexInfinity):
                 raise TypeError("Invalid comparison of complex %s" % me)
             if me is S.NaN:
                 raise TypeError("Invalid NaN comparison")
@@ -269,7 +271,8 @@ def __gt__(self, other):
         except SympifyError:
             raise TypeError("Invalid comparison %s > %s" % (self, other))
         for me in (self, other):
-            if me.is_complex and me.is_real is False:
+            if (me.is_complex and me.is_real is False) or \
+                    me.has(S.ComplexInfinity):
                 raise TypeError("Invalid comparison of complex %s" % me)
             if me is S.NaN:
                 raise TypeError("Invalid NaN comparison")
@@ -286,7 +289,8 @@ def __lt__(self, other):
         except SympifyError:
             raise TypeError("Invalid comparison %s < %s" % (self, other))
         for me in (self, other):
-            if me.is_complex and me.is_real is False:
+            if (me.is_complex and me.is_real is False) or \
+                    me.has(S.ComplexInfinity):
                 raise TypeError("Invalid comparison of complex %s" % me)
             if me is S.NaN:
                 raise TypeError("Invalid NaN comparison")
```

**Comment by aktech** [comment]
> The following passes tests, too.

Have you tried `./bin/doctest integrals` ?
(The Failed case is in Example docs). It Fails on my system.

```
__________________ sympy.integrals.transforms._simplifyconds ___________________
File "sympy/integrals/transforms.py", line 881, in sympy.integrals.transforms._simplifyconds
Failed example:
    simp(abs(x**2) < 1, x, 0)
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest sympy.integrals.transforms._simplifyconds[5]>", line 1, in <module>
        simp(abs(x**2) < 1, x, 0)
      File "sympy/integrals/transforms.py", line 948, in _simplifyconds
        expr = repl(expr, StrictLessThan, replie)
      File "sympy/integrals/transforms.py", line 947, in repl
        return ex.replace(*args)
      File "sympy/core/basic.py", line 1351, in replace
        rv = bottom_up(self, rec_replace, atoms=True)
      File "sympy/simplify/simplify.py", line 4083, in bottom_up
        rv = F(rv)
      File "sympy/core/basic.py", line 1336, in rec_replace
        new = _value(expr, result)
      File "sympy/core/basic.py", line 1280, in <lambda>
        _value = lambda expr, result: value(*expr.args)
      File "sympy/integrals/transforms.py", line 933, in replie
        r = bigger(x, y)
      File "sympy/integrals/transforms.py", line 919, in bigger
        return bigger(1/ex2, 1/ex1)
      File "sympy/integrals/transforms.py", line 925, in bigger
        if n < 0 and (abs(ex1) >= abs(a)**n) == True:
      File "sympy/core/numbers.py", line 1806, in __ge__
        return Rational.__ge__(self, other)
      File "sympy/core/numbers.py", line 1457, in __ge__
        return Expr.__ge__(expr, other)
      File "sympy/core/expr.py", line 240, in __ge__
        raise TypeError("Invalid comparison of complex %s" % me)
    TypeError: Invalid comparison of complex zoo
```

Probably you have tried `sympy/integrals/tests/test_transforms.py` .

File: sympy/core/tests/test_relational.py [hunk 1]
```diff
@@ -459,6 +459,13 @@ def test_nan_complex_inequalities():
         assert_all_ineq_raise_TypeError(r, nan)
 
 
+def test_complex_infinity_inequalities():
+    raises(TypeError, lambda: zoo > 0)
+    raises(TypeError, lambda: zoo >= 0)
+    raises(TypeError, lambda: zoo < 0)
+    raises(TypeError, lambda: zoo <= 0)
+
+
```
Comment (cbm755) [inline sympy/core/tests/test_relational.py]: 1.  suggest `assert_all_ineq_raise_TypeError`: there are lots of paths through this code and I don't trust comparison with zero to capture them all.

2.  Personally, I would do something more like `test_nan_inequality_raise_errors()` above, but @smichr might not agree.

File: sympy/core/tests/test_relational.py [hunk 2]
```diff
@@ -459,6 +459,13 @@ def test_nan_complex_inequalities():
         assert_all_ineq_raise_TypeError(r, nan)
```
Comment (cbm755) [inline sympy/core/tests/test_relational.py]: Maybe add `zoo` to the list on line 458:

````
for r in (I, Symbol('z', imaginary=True), zoo):
````

File: sympy/core/tests/test_relational.py (hunk 1 above)
Comment (cbm755) [inline sympy/core/tests/test_relational.py]: Oops, the "lots of paths" comment goes with 2. not 1.

**Comment by smichr** [comment]
> Have you tried

the diff I gave was on top of your current branch so the catching of the TypeError was not shown in the diff

File: sympy/core/tests/test_relational.py (hunk 2 above)
Comment (aktech) [inline sympy/core/tests/test_relational.py]: I think that would be good, for `nan` comparison.
Thanks!

**Comment by aktech** [comment]
@smichr  
Ah! I see.
Your diff looks better to me, I have made those changes.
Thanks!

**Comment by skirpichev** [comment]
`zoo.is_real is None` - the root of the problem.  If it's true - `zoo > 0` make sense and bug is wrong.  If not, this fix - is not a right one, just an ugly workaround.

**Comment by aktech** [comment]
@skirpichev 
So you mean, we should set `zoo` to be not real in the assumption system.
That would be good I think.
> If not, this fix - is not a right one, just an ugly workaround.

I would be happy to revert this for anything better we could do.

**Pull request merged**